ssl/quic/quic_port.c

```diff
@@ -647,9 +659,126 @@ static int port_try_handle_stateless_reset(QUIC_PORT *port, const QUIC_URXE *e)
     return i > 0;
 }
 
-#define TOKEN_LEN (sizeof("openssltoken") + \
-                   QUIC_RETRY_INTEGRITY_TAG_LEN - 1 + \
-                   sizeof(unsigned char))
+static void cleanup_validation_token(QUIC_VALIDATION_TOKEN *token)
+{
+    OPENSSL_free(token->remote_addr);
+}
+
+/**
+ * @brief Generates a validation token for a RETRY packet.
+ *
+ * @param peer  Address of the client peer receiving the packet.
+ * @param odcid DCID of the connection attempt.
+ * @param rscid Retry source connection ID of the connection attempt.
+ * @param token Address of token to fill data.
+ *
+ * @return 1 if validation token is filled successfully, 0 otherwise.
+ */
+static int generate_retry_token(BIO_ADDR *peer, QUIC_CONN_ID odcid,
+                                QUIC_CONN_ID rscid, QUIC_VALIDATION_TOKEN *token)
+{
+    token->is_retry = 1;
+    token->timestamp = ossl_time_now();
+    token->remote_addr = NULL;
+    token->odcid = odcid;
+    token->rscid = rscid;
+
+    if (!BIO_ADDR_rawaddress(peer, NULL, &token->remote_addr_len)
+        || token->remote_addr_len == 0
+        || (token->remote_addr = OPENSSL_malloc(token->remote_addr_len)) == NULL
+        || !BIO_ADDR_rawaddress(peer, token->remote_addr,
+                                &token->remote_addr_len)) {
+        OPENSSL_free(token->remote_addr);
+        return 0;
+    }
+
+    return 1;
+}
+
+/**
+ * @brief Marshals a validation token into a new buffer.
+ *
+ * Dynamically allocates |buffer| and stores the size of |buffer| in BUFFER_LEN.
+ * Note that it will also allocate an extra QUIC_RETRY_INTEGRITY_TAG_LEN bytes.
+ * The caller is responsible for freeing |buffer|.
+ *
+ * @param token      Validation token.
+ * @param buffer     Callee will allocate a buffer and store the address here.
+ * @param buffer_len Size of |buffer|.
+ */
+static int marshal_validation_token(QUIC_VALIDATION_TOKEN *token,
+                                    unsigned char **buffer, size_t *buffer_len)
+{
+    WPACKET wpkt = {0};
+    BUF_MEM *buf_mem = BUF_MEM_new();
+
+    if (buf_mem == NULL || (token->is_retry != 0 && token->is_retry != 1))
+        return 0;
+
+    if (!WPACKET_init(&wpkt, buf_mem)
+        || !WPACKET_memset(&wpkt, token->is_retry, 1)
+        || !WPACKET_memcpy(&wpkt, &token->timestamp,
+                           sizeof(token->timestamp))
+        || !WPACKET_sub_memcpy_u16(&wpkt, &token->odcid.id, token->odcid.id_len)
+        || !WPACKET_sub_memcpy_u16(&wpkt, &token->rscid.id, token->rscid.id_len)
+        || !WPACKET_sub_memcpy_u16(&wpkt, token->remote_addr, token->remote_addr_len)
```

Comment:
More optimizations would be possible here - we could avoid sending the odcid/rscid at all for the non-RETRY tokens. We could also use WPACKET_sub_memcpy_u8 as none of these fields will be longer than that.